src/gc.c

```diff
@@ -929,103 +1952,314 @@ int64_t diff_gc_total_bytes(void)
 {
-    for(int i=0; i < obj_counts.size; i+=2) {
-        if (obj_counts.table[i+1] != HT_NOTFOUND) {
-            jl_printf(JL_STDERR, "%d ", obj_counts.table[i+1]-1);
-            jl_static_show(JL_STDERR, (jl_value_t*)obj_counts.table[i]);
+    for(int i=0; i < nums.size; i+=2) {
+        if (nums.table[i+1] != HT_NOTFOUND) {
+            void* ty = nums.table[i];
+            int num = (int)nums.table[i+1] - 1;
+            size_t sz = (int)ptrhash_get(&sizes, ty) - 1;
+            jl_printf(JL_STDERR, "   %6d : %4d kB of ", num, sz/1024);
+            if (ty == BUFFTY)
+                jl_printf(JL_STDERR, "buffer");
+            else if (ty == MATY)
+                jl_printf(JL_STDERR, "malloc");
+            else
+                jl_static_show(JL_STDERR, (jl_value_t*)ty);
             jl_printf(JL_STDERR, "\n");
         }
     }
 }
+
+void print_obj_profiles(void)
+{
+    jl_printf(JL_STDERR, "Transient mark :\n");
+    print_obj_profile(obj_counts[0], obj_sizes[0]);
+    jl_printf(JL_STDERR, "Perm mark :\n");
+    print_obj_profile(obj_counts[1], obj_sizes[1]);
+    jl_printf(JL_STDERR, "Remset :\n");
+    print_obj_profile(obj_counts[2], obj_sizes[2]);
+}
 #endif
 
-void jl_gc_collect(void)
+int saved_mark_sp = 0;
+int sweep_mask = GC_MARKED;
+#define MIN_SCAN_BYTES 1024*1024
+
+static void gc_mark_task_stack(jl_task_t*,int);
+
+void prepare_sweep(void)
 {
-    size_t actual_allocd = allocd_bytes;
-    total_allocd_bytes += allocd_bytes;
-    allocd_bytes = 0;
-    if (is_gc_enabled) {
-        JL_SIGATOMIC_BEGIN();
-        jl_in_gc = 1;
-        uint64_t t0 = jl_hrtime();
-        gc_mark();
-#ifdef GCTIME
-        JL_PRINTF(JL_STDERR, "mark time %.3f ms\n", (jl_hrtime()-t0)*1.0e6);
+}
+
+#ifdef GC_VERIFY
+static void clear_mark(int);
 #endif
-#if defined(MEMPROFILE)
-        all_pool_stats();
-        big_obj_stats();
+
+
+void jl_gc_collect(int full)
+{
+    if (!is_gc_enabled) return;
+    if (jl_in_gc) return;
+    jl_in_gc = 1;
+    JL_SIGATOMIC_BEGIN();
+    uint64_t t0 = jl_hrtime();
+    int recollect = 0;
+#if defined(GC_TIME)
+    int wb_activations = mark_sp - saved_mark_sp;
 #endif
-#ifdef GCTIME
-        uint64_t t1 = jl_hrtime();
+    int64_t last_perm_scanned_bytes = perm_scanned_bytes;
+    if (!sweeping) {
+
+        inc_count++;
+        quick_count++;
+
+        scanned_bytes_goal = inc_count*(live_bytes/gc_inc_steps + mark_sp*sizeof(void*));
+        scanned_bytes_goal = scanned_bytes_goal < MIN_SCAN_BYTES ? MIN_SCAN_BYTES : scanned_bytes_goal;
+        if (gc_inc_steps > 1)
+            check_timeout = 1;
+        assert(mark_sp == 0);
+
+        // 1. mark every object in the remset
+        reset_remset();
+        // avoid counting remembered objects & bindings twice in perm_scanned_bytes
+        for(int i = 0; i < last_remset->len; i++) {
+            uintptr_t item = (uintptr_t)last_remset->items[i];
+            void* ptr = (void*)(item & ~(uintptr_t)1);
```

Comment:
I think it can't anymore. I used to use the low bit to denote objects that must be kept in the remset until the next long collection, but it was only used when the bindings didn't have their own remset. I'm pretty sure it's safe to remove this. I can't build right now to confirm.